execution_engine_testing/tests/src/test/chainspec_registry.rs

```diff
@@ -0,0 +1,243 @@
+
+    let (mut builder, _lmdb_fixture_state, _temp_dir) =
+        lmdb_fixture::builder_from_global_state_fixture(lmdb_fixture::RELEASE_1_4_4);
+
+    let mut upgraded_chainspec_registry = ChainspecRegistry::new();
+    let chainspec_bytes_hash = Digest::hash(rng.gen::<[u8; 32]>());
+    upgraded_chainspec_registry.insert(CHAINSPEC_RAW.to_string(), chainspec_bytes_hash);
+    let global_state_toml_hash = Digest::hash(rng.gen::<[u8; 32]>());
+    upgraded_chainspec_registry.insert(GLOBAL_STATE_RAW.to_string(), global_state_toml_hash);
+
+    let mut upgrade_request = {
+        UpgradeRequestBuilder::new()
+            .with_current_protocol_version(*OLD_PROTOCOL_VERSION)
+            .with_new_protocol_version(*NEW_PROTOCOL_VERSION)
+            .with_activation_point(DEFAULT_ACTIVATION_POINT)
+            .with_chainspec_registry(upgraded_chainspec_registry)
+            .build()
+    };
+
+    let engine_config = EngineConfig::default();
+
+    builder
+        .upgrade_with_upgrade_request(engine_config, &mut upgrade_request)
+        .expect_upgrade_success();
+
+    let queried_registry = builder
+        .query(None, Key::ChainspecRegistry, &[])
+        .expect("must have entry under Key::ChainspecRegistry")
+        .as_cl_value()
+        .expect("must have underlying cl_value")
+        .to_owned()
+        .into_t::<ChainspecRegistry>()
+        .expect("must convert to chainspec registry");
+
+    // There should be no entry for the genesis accounts once the upgrade has completed.
+    assert!(queried_registry.get(GENESIS_ACCOUNTS_RAW).is_none());
+
+    let queried_chainspec_hash = queried_registry
+        .get(CHAINSPEC_RAW)
+        .expect("must have entry for chainspec_hash");
+
+    assert_eq!(*queried_chainspec_hash, chainspec_bytes_hash);
+
+    let queried_global_state_toml_hash = queried_registry
+        .get(GLOBAL_STATE_RAW)
+        .expect("must have entry for genesis accounts");
+
+    assert_eq!(*queried_global_state_toml_hash, global_state_toml_hash);
+}
+
+#[ignore]
+#[test]
+fn should_fail_upgrade_when_registry_is_missing_chainspec_hash() {
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&DEFAULT_RUN_GENESIS_REQUEST);
+
+    let invalid_chainspec_registry = ChainspecRegistry::new();
+
+    let mut upgrade_request = {
+        UpgradeRequestBuilder::new()
+            .with_current_protocol_version(*OLD_PROTOCOL_VERSION)
+            .with_new_protocol_version(*NEW_PROTOCOL_VERSION)
+            .with_activation_point(DEFAULT_ACTIVATION_POINT)
+            .with_chainspec_registry(invalid_chainspec_registry)
+            .build()
+    };
+
+    let engine_config = EngineConfig::default();
+
+    builder.upgrade_with_upgrade_request(engine_config, &mut upgrade_request);
+
+    let upgrade_result = builder
+        .get_upgrade_result(0)
+        .expect("must have upgrade result")
+        .clone();
+
+    assert!(matches!(upgrade_result, Err(Error::MissingChainspecHash)))
```

Comment:
Will this work instead?

```
    assert!(matches!(
        upgrade_result,
        Some(Err(Error::MissingChainspecHash))
    ))
```

Reply:
This could work, but I prefer the match on the concrete error, and unwrapping it earlier. Do you think we must implement this for some reason I don't get?

Reply:
That's ok :+1: